youtube_dl/extractor/jiosaavn.py

```diff
@@ -0,0 +1,85 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import random
+import string
+import itertools
+import time
+
+from .common import InfoExtractor
+from ..utils import (
+    int_or_none,
+    urlencode_postdata,
+    unescapeHTML,
+)
+
+
+class JioSaavnIE(InfoExtractor):
+    _VALID_URL = r'''(?x)
+                    https?://(?:www\.)?
+                        (?:
+                            jiosaavn\.com/song/[^/]+/|
+                            saavn.com/s/song/(?:[^/]+/){3}
+                        )
+                        (?P<id>[\w\d]+)
+                   '''
+    _TESTS = [{
+        'url': 'https://www.jiosaavn.com/song/leja-re/OQsEfQFVUXk',
+        'md5': 'd67db3da18f4e72c5558b213d6698704',
+        'info_dict': {
+            'id': 'OQsEfQFVUXk',
+            'ext': 'mp3',
+            'title': 'Leja Re',
+            'album': 'Leja Re',
+            'release_year': 2018,
+        },
+    }, {
+        'url': 'https://www.saavn.com/s/song/hindi/Saathiya/O-Humdum-Suniyo-Re/KAMiazoCblU',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        audio_id = self._match_id(url)
+
+        # A call to this endpoint is a prerequisite for the server to return the auth_url extracted below
+        self._download_webpage('https://www.jiosaavn.com/stats.php', audio_id,
+                               query={
+                                   'ev': 'site:browser:fp',
+                                   'fp': ''.join([random.choice(string.hexdigits) for _ in range(32)]),
+                                   '_t': int(time.time()),
+                                   'ct': '00000000',
+                               })
+
+        webpage = self._download_webpage(url, audio_id)
+
+        song_data = self._parse_json(self._search_regex(r'(?s)<div class="hide song-json">(?P<json>.+?)\</div>',
```

Comment:
Carry arguments to the next lines not to waste space on the left. Keep 80 char len per line.